### DISCUSSION TITLE
Linux build and run instructions

### DISCUSSION THREAD

**Discussion by ShadowMitia** [SebLague/Chess-Challenge#3] [General]
I managed to build and run the program on Linux, I'm sharing the instructions here.
If you have more experience please share it here!

@SebLague Is there a good place to put build instructions? Especially for people who don't use C# on a daily basis? :smile: 
Is the FAQ a good place to do it?

Anyway, here are the steps I used:

- Clone the repo
- Install .NET 6.0 package
There are instructions here: https://learn.microsoft.com/en-us/dotnet/core/install/linux#packages for your OS

Example for people who use `apt` (Debian/Ubuntu and derivatives):

```bash
sudo apt install dotnet-sdk-6.0
```

This should also install the proper runtime.

- Build:

At the root of the repo you can run

```bash
dotnet build
```

You don't need to run this step everytime you make a modification.

- Run:

Go to the directory where the ".csproj" is and do `dotnet run`.

```bash
cd Chess-Challenge
dotnet run
```

Alternatively, go to where the executable is created and run it directly.

```bash
cd Chess-Challenge/bin/Debug/net6.0/
dotnet Chess-Challenge.dll
```

If everything worked well, you should be able to see the chessboard and play with it :smiley: 

Thanks to @JoseBritto for the addendums on building and running!

**Comment by Aaron-Speedy** [comment]
For anyone trying to run this on Arch, make sure you install ``dotnet-sdk-6.0`` rather than ``dotnet-sdk``.

**Comment by JoseBritto** [comment]
> 
>    Run:
>
>It seems you have to be in the same folder as the executable?
>```
>cd Chess-Challenge/bin/Debug/net6.0/
>dotnet Chess-Challenge.dll
>```

Or use `dotnet run` after cd ing into the directory with the `Chess-Challenge.csproj` file. 
You don't have to do `dotnet build` everytime you want to run the project after modifying something.

**Reply by ShadowMitia** [reply]
Good to know! I'll update my instructions, thanks!

**Comment by SebLague** [comment]
Thanks for sharing! I've added a link to this discussion to the readme.

**Comment by ElphaTech** [comment]
I followed your instructions but when running `dotnet Chess-Challenge.dll` it gives the following error:
```
Unhandled exception. System.DllNotFoundException: Unable to load shared library 'raylib' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libraylib: cannot open shared object file: No such file or directory
   at Raylib_cs.Raylib.SetTraceLogCallback(IntPtr callback)
   at ChessChallenge.Application.Program.Main() in /home/myName/Desktop/ChessBotChallange/Chess-Challenge/src/Framework/Application/Core/Program.cs:line 23
Aborted (core dumped)```
Do you know how to fix it. Sorry if this is a rookie error, I have never used anything like this before.

**Reply by ShadowMitia** [reply]
No idea how to fix this, but I've seen other people have issues with raylib, you should probably check out the Issues.
Can you give more details on how you tried to run? What OS are you running on?

**Reply by Vitaspiros** [reply]
I have had the same problem. To fix it, I followed [this](https://github.com/raysan5/raylib/wiki/Working-on-GNU-Linux) page for how to build raylib and install it. After I did that, I am presented with the following error:
```
dotnet: symbol lookup error: /opt/amdgpu/lib/x86_64-linux-gnu/libGLX_mesa.so.0: undefined symbol: drmGetDevice2
```
It seems to have to do with my graphics driver since it mentions `amdgpu`. Can you help  me?

**Reply by ElphaTech** [reply]
I'm using Ubuntu 20.04. I just ran it the way directed above using terminal.

**Reply by Zokaper** [reply]
I'm also having this issue. Pop!_OS, just followed the post's instructions. I have also tried installing raylib to no avail.

**Reply by guillaumecarraux** [reply]
I solved it by copying the file `libraylib.so` from /usr/local/lib to bin/Debug/net6.0 directory

**Reply by carsonetb** [reply]
Copying the libraylib.so file to net6.0 directory did not work for me (Debian)

**Reply by guillaumecarraux** [reply]
I am also Debian, can you send the exact error message that you get ?

**Reply by kabrownie** [reply]
Iam getting the same error

**Reply by kabrownie** [reply]
Unhandled exception. System.DllNotFoundException: Unable to load shared library 'raylib' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libraylib: cannot open shared object file: No such file or directory
   at Raylib_cs.Raylib.SetTraceLogCallback(IntPtr callback)
   at ChessChallenge.Application.Program.Main() in /home/kabrownie/projects/Chess-Challenge/Chess-Challenge/src/Framework/Application/Core/Program.cs:line 23

**Reply by pascscha** [reply]
My issue was that the raylib version I installed was using `GLIBC_2.35`, while I had version 2.31 (as shown by `ldd --version`). [Building raylib from source using cmake did the trick for me.](https://github.com/raysan5/raylib/wiki/Working-on-GNU-Linux#build-raylib-using-cmake)

**Reply by SquarishRectangle** [reply]
Some older LTS versions of Ubuntu are using outdated versions of GLIBC. Upgrading to 22.04 worked for me.

**Reply by Saggre** [reply]
Works after building `libraylib.so` from [source](https://github.com/raysan5/raylib/wiki/Working-on-GNU-Linux) and moving it to `bin/Debug/net6.0`

**Reply by Joshua-Moler** [reply]
Copying libraylib.so to /bin/Debug/net6.0 worked on Mint 20.3.

**Comment by Vilsol** [comment]
I am attempting to run this on NixOS 23.05. I have installed the `dotnet-sdk_6` package. I am able to build the project just fine, however, the issue is running it. I seem to be getting a segfault. All steps executed:
```shell
⛵kubernetes-ab ~/Projects/SebLague 
[10:17:47] ❯ git clone git@github.com:SebLague/Chess-Challenge.git
Cloning into 'Chess-Challenge'...
remote: Enumerating objects: 152, done.
remote: Counting objects: 100% (45/45), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 152 (delta 22), reused 20 (delta 20), pack-reused 107
Receiving objects: 100% (152/152), 3.39 MiB | 4.75 MiB/s, done.
Resolving deltas: 100% (41/41), done.

⛵kubernetes-ab ~/Projects/SebLague took 4s 
[10:17:56] ❯ cd Chess-Challenge                  

⛵kubernetes-ab Chess-Challenge on  main 
[10:18:00] ❯ dotnet build
MSBuild version 17.3.2+561848881 for .NET
  Determining projects to restore...
  Restored /home/vilsol/Projects/SebLague/Chess-Challenge/Chess-Challenge/Chess-Challenge.csproj (in 241 ms).
/home/vilsol/Projects/SebLague/Chess-Challenge/Chess-Challenge/src/Framework/Chess/Board/Board.cs(71,37): warning CS8625: Cannot convert null literal to non-nullable reference type. [/home/vilsol/Projects/SebLague/Chess-Challenge/Chess-Challenge/Chess-Challenge.csproj]
  Chess-Challenge -> /home/vilsol/Projects/SebLague/Chess-Challenge/Chess-Challenge/bin/Debug/net6.0/Chess-Challenge.dll

Build succeeded.

/home/vilsol/Projects/SebLague/Chess-Challenge/Chess-Challenge/src/Framework/Chess/Board/Board.cs(71,37): warning CS8625: Cannot convert null literal to non-nullable reference type. [/home/vilsol/Projects/SebLague/Chess-Challenge/Chess-Challenge/Chess-Challenge.csproj]
    1 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.53

⛵kubernetes-ab Chess-Challenge on  main took 2s 
[10:18:04] ❯ cd Chess-Challenge/bin/Debug/net6.0 

⛵kubernetes-ab bin/Debug/net6.0 on  main 
[10:18:25] ❯ dotnet Chess-Challenge.dll     
Launching Chess-Challenge version 1.1
[1]    2611399 segmentation fault (core dumped)  dotnet Chess-Challenge.dll
```

**Reply by ShadowMitia** [reply]
Don't know NixOS but maybe you're missing some runtime packages?
Can you get a stacktrace of the segfault?

**Reply by metiulekm** [reply]
I got a similar issue, here's my backtrace:
```
#0  0x0000000000000000 in ?? ()
#1  0x00007f66dc5dce69 in rlLoadTexture () from /tmp/Chess-Challenge/Chess-Challenge/bin/Debug/net6.0/runtimes/linux-x64/native/libraylib.so
#2  0x00007f66dc6694e1 in LoadTextureFromImage () from /tmp/Chess-Challenge/Chess-Challenge/bin/Debug/net6.0/runtimes/linux-x64/native/libraylib.so
#3  0x00007f6669b91379 in ?? ()
#4  0x00007f66cdd42010 in ?? ()
#5  0x0000040000000400 in ?? ()
#6  0x0000000200000001 in ?? ()
#7  0x00000000004aa412 in ?? ()
#8  0x00007f66e2ace8f0 in ?? () from /nix/store/cp3rvj1xy2dvv60l5dk6q925ld15x6wp-dotnet-sdk-6.0.411/shared/Microsoft.NETCore.App/6.0.19/libcoreclr.so
#9  0x00007ffd008d78d8 in ?? ()
#10 0x00007f666948f520 in ?? ()
#11 0x00007f666948f520 in ?? ()
#12 0x00007ffd008d6f20 in ?? ()
#13 0x00007f6669b91379 in ?? ()
#14 0x00007ffd008d7000 in ?? ()
#15 0x0000000200000001 in ?? ()
#16 0x00007f666948f520 in ?? ()
#17 0x00007ffd008d7070 in ?? ()
#18 0x0000001900000052 in ?? ()
#19 0x000001f600000022 in ?? ()
#20 0x00007f66cdd42010 in ?? ()
#21 0x00007ffd008d7070 in ?? ()
#22 0x0000000000000000 in ?? ()
```
It seems that GL function loading is not working somehow…?

**Reply by metiulekm** [reply]
Actually, I just got it working, albeit in a very hacky way. @Vilsol, try this:
```sh
cp $(nix-build '<nixpkgs>' -A raylib --no-link)/lib/libraylib.so $HOME/.nuget/packages/raylib-cs/4.5.0.2/runtimes/linux-x64/native/libraylib.so
```
and then you might need to launch `dotnet run`/`dotnet Chess-Challenge.dll` via `nixGL`.

**Reply by Vilsol** [reply]
@metiulekm indeed works with that magic copy command. I know absolutely nothing about the dotnet package management, but I'm guessing it's not searching the full path for the `raylib`? As I tried installing it system wide, but that did not fix it. I also did not need to use `nixGL`

**Reply by CactiChameleon9** [reply]
@metiulekm Just wanted to thank you for this solution! I have no clue why it works, but helpful nevertheless

**Comment by sanchit1053** [comment]
I am new to CSharp and dotnet development, Is the version 6.0 a hard requirement for the program to work?

**Reply by DaveAxiom** [reply]
Yes. Version 6 of [Microsoft DotNet Core](https://packages.ubuntu.com/search?keywords=dotnet-sdk-6.0) is required.

I tried [Mono](https://www.mono-project.com/) and found within [MonoDevelop](https://www.monodevelop.com/) compilation errors reporting that Mono (or Monodevelop) supports only up to v.4.6. The Mono website claims there is v.6 support but I'm not finding that to be the case when compiling. If Mono has proper support, it would take only the correct command lines to compile outside of MonoDevelop. Either MonoDevelop (or Mono) needs work or Chess-Challenge itself needs to be back ported to work with only Free Software.

**Reply by fuj1n** [reply]
.NET 6 is free (and open), so I'm not really sure what you mean.

**Reply by gsckoco** [reply]
> I tried [Mono](https://www.mono-project.com/) and found within [MonoDevelop](https://www.monodevelop.com/) compilation errors reporting that Mono (or Monodevelop) supports only up to v.4.6. The Mono website claims there is v.6 support but I'm not finding that to be the case when compiling. If Mono has proper support, it would take only the correct command lines to compile outside of MonoDevelop. Either MonoDevelop (or Mono) needs work or Chess-Challenge itself needs to be back ported to work with only Free Software.

Why would you expect mono to work? Mono is .NET Framework and not .NET Core. .NET Core is open source and cross platform, unlike what Mono is trying to achieve for .NET Framework.

**Comment by AlbinoElk** [comment]
Im using Linux Mint and I get `
Launching Chess-Challenge version 1.13
Segmentation fault (core dumped)
` everytime I try to run it using `dotnet Chess-Challenge.dll`, but if I run it using `dotnet run` I get `Launching Chess-Challenge version 1.13 ` and nothing happens

**Reply by the4thCuber** [reply]
Same

**Comment by dudecon** [comment]
Thanks for the instructions! For those who want to use an IDE, I got it working in Visual Studio Code. Be sure to install a non-flatpack version so that it's not sandboxed.
With VSC open, select:
`File > Open Workspace from file...` and select your "Chess-Challenge/Chess-Challenge.code-workspace" file
and then select:
`Run > Add Configuration...` and select `.NET 5+ and .NET Core` from the drop-down
This will add a couple .json files so that the IDE knows how to run the code. Hopefully it will just run from that point.

**Reply by ElphaTech** [reply]
Is there a way to check if it is not sandboxed?

**Reply by dudecon** [reply]
I'm not sure. I just know that the flathub (flatpack) install of Visual Studio Code in my Ubuntu flavored distro couldn't see the .NET 6.0 libraries, and the Pop!_OS (deb) version works just fine.

**Reply by fuj1n** [reply]
> Is there a way to check if it is not sandboxed?

If you installed it from Flatpak (or snap, not sure if there is a snap version), it is sandboxed.

Flatpak expressly prohibits access to system files, so I think if you do `dotnet run /etc`, it should give you an error about not being able to access that folder if it is sandboxed.

**Comment by endrawes0** [comment]
These steps worked for me on Mac as well! Thank you!

**Reply by endrawes0** [reply]
Well, I should be more clean, VS did not, for whatever reason recognize the .NET6 installation. But I was able to build it from the command line - this idea was inspired by this post. 

Whatever means of installing the .NET6 SDK should be sufficient to replace apt get above, I think.

**Comment by jackhassenzahl** [comment]
trying to make this work on a raspberry pi 5 but I'm struggling. I just want to have a go at making a bot with these and try against the other bots that were in the challenge. if it helps I'm using Debian 12 and making a fresh OS image is not an issue if I have to.

**Reply by Saggre** [reply]
What kind of problem did you encounter?

**Reply by jackhassenzahl** [reply]
When I tried to run the code via terminal I hade an error that said I was missing libraylib and I've tried everything to install it. It could be in the wrong path when I tried to install it. I'm using Ubuntu 23.10 and I'm wondering if that package is supposed to be there but just isn't.

**Reply by jackhassenzahl** [reply]
When I try to run it with VScode, I get an error that says I need an XML debugger. I saw you have to make a debug configuration. I selected .NET and then I can choose to launch with .NET or do something else with C#. Launch with .NET just seems to load forever and the C# option throws a workspace error. I've never used .NET or C# and I'm relatively new to raspberry pi so this is all new to me.

**Reply by Saggre** [reply]
@ISaidJak Build libraylib from source on the actual raspberry machine: https://github.com/raysan5/raylib/wiki/Working-on-GNU-Linux
Then move the built library file to the project's bin/Debug/net6.0

**Reply by Saggre** [reply]
Before building you can also try moving any of the libraries you already have to bin/Debug/net6.0

**Reply by jackhassenzahl** [reply]
What built files do i need to move? The whole folder or just some files? I moved 3 libraylib.so files and I stopped getting errors but nothing happens
